<a href="https://colab.research.google.com/github/abilashkannanv/AIML/blob/main/TrustCart_Phase1_Transaction_Risk_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

## Task 1.1 Dataset Familiarization

In [2]:
# Load transactions.csv
transactions_df = pd.read_csv('/content/sample_data/transaction.csv')

# Load identity.csv
identity_df = pd.read_csv('/content/sample_data/identity.csv')

print('Datasets loaded successfully.')

Datasets loaded successfully.


### Analysis of `transactions.csv`

In [3]:
# Identify number of records and features for transactions_df
num_records_transactions = transactions_df.shape[0]
num_features_transactions = transactions_df.shape[1]

print(f"Number of records in transactions.csv: {num_records_transactions}")
print(f"Number of features in transactions.csv: {num_features_transactions}")

# Check for target variable (risk_flag) in transactions_df
if 'risk_flag' in transactions_df.columns:
    print("Target variable 'risk_flag' found in transactions.csv.")
else:
    print("Target variable 'risk_flag' not found in transactions.csv.")

print("\nFirst 5 rows of transactions.csv:")
display(transactions_df.head())

print("\nInformation about transactions.csv:")
transactions_df.info()

Number of records in transactions.csv: 296234
Number of features in transactions.csv: 394
Target variable 'risk_flag' not found in transactions.csv.

First 5 rows of transactions.csv:


,TransactionID,isFraud,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,...,V330,V331,V332,V333,V334,V335,V336,V337,V338,V339
0,2987000,0,86400,68.5,W,13926,NaN,150.0,discover,142.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2987001,0,86401,29.0,W,2755,404.0,150.0,mastercard,102.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2987002,0,86469,59.0,W,4663,490.0,150.0,visa,166.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2987003,0,86499,50.0,W,18132,567.0,150.0,mastercard,117.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2987004,0,86506,50.0,H,4497,514.0,150.0,mastercard,102.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0



Information about transactions.csv:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 296234 entries, 0 to 296233
Columns: 394 entries, TransactionID to V339
dtypes: float64(376), int64(4), object(14)
memory usage: 890.5+ MB


### Analysis of `identity.csv`

In [4]:
# Identify number of records and features for identity_df
num_records_identity = identity_df.shape[0]
num_features_identity = identity_df.shape[1]

print(f"Number of records in identity.csv: {num_records_identity}")
print(f"Number of features in identity.csv: {num_features_identity}")

# Check for target variable (risk_flag) in identity_df
if 'risk_flag' in identity_df.columns:
    print("Target variable 'risk_flag' found in identity.csv.")
else:
    print("Target variable 'risk_flag' not found in identity.csv.")

print("\nFirst 5 rows of identity.csv:")
display(identity_df.head())

print("\nInformation about identity.csv:")
identity_df.info()

Number of records in identity.csv: 144233
Number of features in identity.csv: 41
Target variable 'risk_flag' not found in identity.csv.

First 5 rows of identity.csv:


,TransactionID,id_01,id_02,id_03,id_04,id_05,id_06,id_07,id_08,id_09,...,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,2987004,0.0,70787.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,samsung browser 6.2,32.0,2220x1080,match_status:2,T,F,T,T,mobile,SAMSUNG SM-G892A Build/NRD90M
1,2987008,-5.0,98945.0,NaN,NaN,0.0,-5.0,NaN,NaN,NaN,...,mobile safari 11.0,32.0,1334x750,match_status:1,T,F,F,T,mobile,iOS Device
2,2987010,-5.0,191631.0,0.0,0.0,0.0,0.0,NaN,NaN,0.0,...,chrome 62.0,NaN,NaN,NaN,F,F,T,T,desktop,Windows
3,2987011,-5.0,221832.0,NaN,NaN,0.0,-6.0,NaN,NaN,NaN,...,chrome 62.0,NaN,NaN,NaN,F,F,T,T,desktop,NaN
4,2987016,0.0,7460.0,0.0,0.0,1.0,0.0,NaN,NaN,0.0,...,chrome 62.0,24.0,1280x800,match_status:2,T,F,T,T,desktop,MacOS



Information about identity.csv:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144233 entries, 0 to 144232
Data columns (total 41 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionID  144233 non-null  int64  
 1   id_01          144233 non-null  float64
 2   id_02          140872 non-null  float64
 3   id_03          66324 non-null   float64
 4   id_04          66324 non-null   float64
 5   id_05          136865 non-null  float64
 6   id_06          136865 non-null  float64
 7   id_07          5155 non-null    float64
 8   id_08          5155 non-null    float64
 9   id_09          74926 non-null   float64
 10  id_10          74926 non-null   float64
 11  id_11          140978 non-null  float64
 12  id_12          144233 non-null  object 
 13  id_13          127320 non-null  float64
 14  id_14          80044 non-null   float64
 15  id_15          140985 non-null  object 
 16  id_16          129340 non-null  object 
 

### Business Meaning of Columns
To fully understand the business meaning of columns, a data dictionary is required. Since it's not provided, I will proceed with the assumption that `risk_flag` is the target variable based on the instructions. The `df.info()` output above provides data types and non-null counts, which helps in initial understanding of the column characteristics.

From the head and info of both dataframes, we can infer some column meanings:

**`transactions.csv`**:
*   `TransactionID`: Unique identifier for each transaction.
*   `isFraud`: Likely indicates if a transaction is fraudulent (this would be a target variable if not `risk_flag`).
*   `TransactionDT`: Timestamp of the transaction.
*   `TransactionAmt`: Amount of the transaction.
*   `ProductCD`: Product code.
*   `card1` to `card6`: Anonymized card features.
*   `addr1`, `addr2`: Anonymized address features.
*   `P_emaildomain`, `R_emaildomain`: Primary and secondary email domains.
*   `C1` to `C14`: Anonymized counting features.
*   `D1` to `D15`: Anonymized timedelta features.
*   `M1` to `M9`: Anonymized matching features.
*   `V1` to `V339`: Anonymized Vesta engineered features.

**`identity.csv`**:
*   `TransactionID`: Unique identifier for each transaction (common key).
*   `DeviceType`, `DeviceInfo`: Information about the device used for the transaction.
*   `id_01` to `id_38`: Anonymized identity features.

**Target Variable:** As per the instructions, `risk_flag` is the target variable. I will assume this is present in one of the datasets or will be derived. Based on the `transactions.csv` content, `isFraud` seems like the most likely candidate for `risk_flag` if it's not explicitly named `risk_flag`.

## Task 1.2 Data Joining Strategy

### Identify Common Keys and Join Type

**Common Key:**
Both `transactions_df` and `identity_df` contain a column named `TransactionID`. This column appears to be a unique identifier for each transaction and is present in both datasets, making it the ideal common key for merging.

**Join Type and Justification:**

I will use a **left join** (or `merge` with `how='left'`) to combine `identity_df` with `transactions_df`.

*   **`transactions_df` (left dataframe):** Contains 296,234 records, representing the main set of transactions.
*   **`identity_df` (right dataframe):** Contains 144,233 records, providing additional identity-related information for a subset of the transactions.

**Justification for Left Join:**

1.  **Retain all transaction records:** The primary goal is often to analyze all transactions, whether they have associated identity information or not. A left join ensures that all records from `transactions_df` are kept.
2.  **Enrichment:** Identity information from `identity_df` will be added to the matching transaction records. For transactions present in `transactions_df` but not in `identity_df`, the identity-related columns will have `NaN` (Not a Number) values, indicating missing identity information.
3.  **Preservation of context:** Since `transactions_df` is larger and likely represents the core dataset for fraud detection, keeping all its records is crucial for a comprehensive analysis.

This strategy allows us to analyze the impact of identity features on all transactions, while also being able to identify transactions where identity data is unavailable.

In [5]:
# Perform a left merge
merged_df = pd.merge(transactions_df, identity_df, on='TransactionID', how='left')

print("Merged DataFrame created successfully.")
print(f"Number of records in merged_df: {merged_df.shape[0]}")
print(f"Number of features in merged_df: {merged_df.shape[1]}")

print("\nFirst 5 rows of merged_df:")
display(merged_df.head())

print("\nInformation about merged_df (showing first few columns):")
merged_df.info(verbose=True, show_counts=True, max_cols=50)

Merged DataFrame created successfully.
Number of records in merged_df: 296234
Number of features in merged_df: 434

First 5 rows of merged_df:


,TransactionID,isFraud,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,...,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,2987000,0,86400,68.5,W,13926,NaN,150.0,discover,142.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2987001,0,86401,29.0,W,2755,404.0,150.0,mastercard,102.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2987002,0,86469,59.0,W,4663,490.0,150.0,visa,166.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2987003,0,86499,50.0,W,18132,567.0,150.0,mastercard,117.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2987004,0,86506,50.0,H,4497,514.0,150.0,mastercard,102.0,...,samsung browser 6.2,32.0,2220x1080,match_status:2,T,F,T,T,mobile,SAMSUNG SM-G892A Build/NRD90M



Information about merged_df (showing first few columns):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 296234 entries, 0 to 296233
Data columns (total 434 columns):
 #    Column          Non-Null Count   Dtype  
---   ------          --------------   -----  
 0    TransactionID   296234 non-null  int64  
 1    isFraud         296234 non-null  int64  
 2    TransactionDT   296234 non-null  int64  
 3    TransactionAmt  296234 non-null  float64
 4    ProductCD       296234 non-null  object 
 5    card1           296234 non-null  int64  
 6    card2           291520 non-null  float64
 7    card3           296229 non-null  float64
 8    card4           296222 non-null  object 
 9    card5           294641 non-null  float64
 10   card6           296225 non-null  object 
 11   addr1           263339 non-null  float64
 12   addr2           263339 non-null  float64
 13   dist1           111361 non-null  float64
 14   dist2           23648 non-null   float64
 15   P_emaildomain   249473 no